cmd/k8s-operator/proxygroup_specs.go

```diff
@@ -0,0 +1,242 @@
+									mounts = append(mounts, corev1.VolumeMount{
+										Name:      fmt.Sprintf("tailscaledconfig-%d", i),
+										ReadOnly:  true,
+										MountPath: fmt.Sprintf("/etc/tsconfig/%s-%d", pg.Name, i),
+									})
+								}
+
+								return mounts
+							}(),
+						},
+					},
+					Volumes: func() []corev1.Volume {
+						var volumes []corev1.Volume
+						for i := range pgReplicas(pg) {
+							volumes = append(volumes, corev1.Volume{
+								Name: fmt.Sprintf("tailscaledconfig-%d", i),
+								VolumeSource: corev1.VolumeSource{
+									Secret: &corev1.SecretVolumeSource{
+										SecretName: fmt.Sprintf("%s-%d-config", pg.Name, i),
+									},
+								},
+							})
+						}
+
+						return volumes
+					}(),
+				},
+			},
+		},
+	}
+}
+
+func pgServiceAccount(pg *tsapi.ProxyGroup, namespace string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            pg.Name,
+			Namespace:       namespace,
+			Labels:          labels("proxygroup", pg.Name, nil),
+			OwnerReferences: pgOwnerReference(pg),
+		},
+	}
+}
+
+func pgRole(pg *tsapi.ProxyGroup, namespace string) *rbacv1.Role {
+	return &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            pg.Name,
+			Namespace:       namespace,
+			Labels:          labels("proxygroup", pg.Name, nil),
+			OwnerReferences: pgOwnerReference(pg),
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"secrets"},
+				Verbs: []string{
+					"get",
+					"patch",
+					"update",
+				},
+				ResourceNames: func() (secrets []string) {
+					for i := range pgReplicas(pg) {
+						secrets = append(secrets,
+							fmt.Sprintf("%s-%d-config", pg.Name, i), // Config with auth key.
+							fmt.Sprintf("%s-%d", pg.Name, i),        // State.
+						)
+					}
+					return secrets
+				}(),
+			},
+		},
+	}
+}
+
+func pgRoleBinding(pg *tsapi.ProxyGroup, namespace string) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            pg.Name,
+			Namespace:       namespace,
+			Labels:          labels("proxygroup", pg.Name, nil),
+			OwnerReferences: pgOwnerReference(pg),
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      pg.Name,
+				Namespace: namespace,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind: "Role",
+			Name: pg.Name,
+		},
+	}
+}
+
+func pgStateSecrets(pg *tsapi.ProxyGroup, namespace string) (secrets []*corev1.Secret) {
+	for i := range pgReplicas(pg) {
+		secrets = append(secrets, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            fmt.Sprintf("%s-%d", pg.Name, i),
+				Namespace:       namespace,
+				Labels:          secretLabels("proxygroup", pg.Name, "state"),
+				OwnerReferences: pgOwnerReference(pg),
+			},
+		})
+	}
+
+	return secrets
+}
+
+func secretLabels(app, instance, typ string) map[string]string {
+	return labels(app, instance, map[string]string{
```

Comment:
I think we should try to make this more similar to how the existing proxies work:
- label them with `tailscale.com` labels
- use `tailscale.com` labels for lookup
This is for consistency with the existing proxies, because those labels are more semantically meaningful and because with those we can be most certain that there won't be any surprises (users also created a `Secret` with that label for a different purpose and we are now picking that up

Reply:
For resources created for `ProxyGroup` there would be good to have:
- `tailscale.com/managed: true` - we set that for all other proxy resources
- `tailscale.com/proxy-group: <proxy-group-name>` - for lookup

Reply:
Also would be nice to have a separate `proxyGroupChildResourceLabels` function that takes a `ProxyGroup` so it's clearer exactly what labels all the resources created for a `ProxyGroup` will have and what the values are for a specific `ProxyGroup` will be

Reply:
Last bit - managed labels need to be added to https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/sts.go#L101

Reply:
👍 thanks for the notes, added in 101bd89efdab7da6f6fb0488c6ced66c0faf68b7